File: src/main/java/com/college/hotlittlepigs/security/JwtManager.java
package com.college.hotlittlepigs.security;

import com.college.hotlittlepigs.user.dto.UserLoginResponseDTO;
import com.college.hotlittlepigs.user.User;
import io.jsonwebtoken.Claims;
import io.jsonwebtoken.JwtException;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.SignatureAlgorithm;
import org.springframework.stereotype.Component;

import java.util.Calendar;
import java.util.List;

@Component
public class JwtManager {

  public UserLoginResponseDTO createToken(User user, List<String> roles) {
    var calendar = Calendar.getInstance();
    calendar.add(Calendar.DAY_OF_MONTH, SecurityConstant.JWT_EXP_DAYS);

    var jwt =
        Jwts.builder()
            .setSubject(user.getEmail())
            .setExpiration(calendar.getTime())
            .claim(SecurityConstant.JWT_ROLE_KEY, roles)
            .claim("id", user.getId())
            .signWith(SignatureAlgorithm.HS512, SecurityConstant.API_KEY.getBytes())
            .compact();

    Long expireIn = calendar.getTimeInMillis();

    return new UserLoginResponseDTO(jwt, expireIn, SecurityConstant.JWT_PROVIDER);
  }

  public Claims parseToken(String jwt) throws JwtException {
    return Jwts.parser()
        .setSigningKey(SecurityConstant.API_KEY.getBytes())
        .parseClaimsJws(jwt)
        .getBody();
  }
}

File: src/main/kotlin/com/college/hotlittlepigs/security/JwtManager.kt
package com.college.hotlittlepigs.security

import com.college.hotlittlepigs.user.dto.UserLoginResponseDTO
import java.util.Calendar
import io.jsonwebtoken.Jwts
import kotlin.Throws
import io.jsonwebtoken.Claims
import io.jsonwebtoken.JwtException
import io.jsonwebtoken.SignatureAlgorithm
import org.springframework.stereotype.Component

@Component
class JwtManager {
    fun createToken(email: String, roles: List<String>): UserLoginResponseDTO {
        val expiration = Calendar.getInstance()
        expiration.add(Calendar.DAY_OF_MONTH, SecurityConstants.JWT_EXP_DAYS)

        val jwt = Jwts.builder()
            .setSubject(email)
            .setExpiration(expiration.time)
            .claim(SecurityConstants.JWT_ROLE_KEY, roles)
            .signWith(SignatureAlgorithm.HS512, SecurityConstants.API_KEY.toByteArray())
            .compact()

        val expireIn = expiration.timeInMillis
        return UserLoginResponseDTO(jwt, expireIn, SecurityConstants.JWT_PROVIDER)
    }

    @Throws(JwtException::class)
    fun parseToken(jwt: String): Claims = Jwts.parser()
        .setSigningKey(SecurityConstants.API_KEY.toByteArray())
        .parseClaimsJws(jwt)
        .body
}